p-ata/src/entrypoint.rs

```diff
@@ -0,0 +1,126 @@
+#![allow(unexpected_cfgs)]
+
+use {
+    crate::processor::process_create_associated_token_account,
+    crate::recover::process_recover_nested,
+    pinocchio::{
+        account_info::AccountInfo, no_allocator, nostd_panic_handler, program_entrypoint,
+        program_error::ProgramError, pubkey::Pubkey, ProgramResult,
+    },
+};
+
+/// Maximum allowed account `known_account_data_length` to prevent creation
+/// of accounts that are expensive to work with. This mitigates potential
+/// attacks where callers could create very large, expensive-to-work-with
+/// accounts for others.
+///
+/// Any token program requiring a larger length must rely on CPI calls to
+/// determine the account length and cannot pass in `known_account_data_length`.
+pub const MAX_SANE_ACCOUNT_LENGTH: u16 = 1024;
+
+program_entrypoint!(process_instruction);
+no_allocator!();
+nostd_panic_handler!();
+
+/// Main instruction processor for the p-ATA program.
+///
+/// ## Instruction Format
+///
+/// ### Create ATA (Non-Idempotent) - Discriminator: 0 or Empty
+/// ```ignore
+/// [0] or []                     -> compute bump and ATA account data length on-chain  
+/// [0, bump]                     -> use provided bump, compute ATA account data length
+/// [0, bump, len_low, len_high]  -> use provided bump and ATA account data length
+/// ```
+///
+/// ### Create ATA (Idempotent) - Discriminator: 1  
+/// ```ignore
+/// [1]                           -> compute bump and ATA account data length on-chain
+/// [1, bump]                     -> use provided bump, compute ATA account data length  
+/// [1, bump, len_low, len_high]  -> use provided bump and ATA account data length
+/// ```
+///
+/// ### Recover Nested ATA - Discriminator: 2
+/// ```ignore
+/// [2]                                      -> compute all bumps on-chain
+/// [2, owner_bump, nested_bump, dest_bump]  -> use provided bumps
+/// ```
+///
+/// ## Account Layout (Create)
+/// ```ignore
+/// [0] payer                    (signer, writable) - pays for account creation (and rent if applicable)
+/// [1] associated_token_account (writable)         - account to create  
+/// [2] wallet                   (signer)           - token account owner
+/// [3] mint                                        - token mint account
+/// [4] system_program                              - system program
+/// [5] token_program                               - token program
+/// [6] rent_sysvar              (optional)         - rent sysvar (for optimization)
```

Comment:
This was normally only used by the "deprecated" instruction that provides no data, but we can add it back in if it's an optimization!